bot/src/main/java/com/community/tools/service/TaskStatusService.java

```diff
@@ -1,172 +1,82 @@
 
 @Slf4j
-@Service
+@Service("taskStatusServiceHooks")
 public class TaskStatusService {
 
-  private final ClassroomService classroomService;
+  private final UserTaskRepository userTaskRepository;
+  @Value("${tasksForUsers}")
+  private String originalTaskNames;
 
-  public TaskStatusService(ClassroomService classroomService) {
-    this.classroomService = classroomService;
+  public TaskStatusService(UserTaskRepository userTaskRepository) {
+    this.userTaskRepository = userTaskRepository;
   }
 
   /**
    * Service for sorting, limiting and creating DTO.
-   * @param limit - limit of users for view.
-   * @param period - period of days fow view.
+   *
+   * @param limit      - limit of users for view.
+   * @param period     - period of days fow view.
    * @param comparator - sort order (DESC, ASC).
    * @return - return list of DTO.
    */
   public List<UserForTaskStatusDto> getTaskStatuses(Period period, Integer limit,
-      Comparator<GithubUserDto> comparator) {
-    log.info("running with period = {}, comparator = {}, limit = {}", period, comparator, limit);
-    return classroomService.getAllActiveUsers(period).stream()
-        .filter(this::isActiveUser)
+      Comparator<UserForTaskStatusDto> comparator) {
+    LocalDate startDate = LocalDate.ofEpochDay(LocalDate
+        .now()
+        .minus(period).toEpochDay());
+    return userTaskRepository
+        .findUserTasksByLastActivityAfter(startDate)
+        .stream()
+        .collect(Collectors.groupingBy(UserTask::getGitName))
+        .entrySet()
+        .stream()
+        .map(entry -> {
+          List<TaskNameAndStatus> taskNamesAndStatuses = new ArrayList<>();
+          entry.getValue().forEach(
+              task -> taskNamesAndStatuses.add(
+                  new TaskNameAndStatus(task.getTaskName(), task.getPullUrl(),
+                      task.getTaskStatus())));
+
+          final int completedTasksCount = (int) taskNamesAndStatuses.stream()
+              .filter(it -> originalTaskNames.contains(it.getTaskName()))
+              .filter(it -> it.getTaskStatus().equals(
+                  TaskStatus.DONE.getDescription())).count();
+
+          /* This is supposed to throw NoSuchElementException
+          in case it cannot find lastActivity for gitName
+          because it is PPK in db, therefore at least one such record MUST exist,
+          if it does not, something went terribly wrong,
+          there is no point in trying to recover.
+           */
+          try {
+            final LocalDate lastActive = entry.getValue().stream().max(
+                Comparator.comparing(UserTask::getLastActivity)).get().getLastActivity();
+            return new UserForTaskStatusDto(entry.getKey(),
+                lastActive, completedTasksCount, taskNamesAndStatuses);
+          } catch (NoSuchElementException e) {
+            log.error("Could not find lastActive time for user " + entry.getKey(), e);
+            throw e;
```

Comment:
I see no reason to rethrow this. You don't have a handler upstream, so this basically creates two same exceptions in the stacktrace.

Reply:
The map Function needs to return ot throw something, at this stage any DTO that could possibly be returned will be malformed, so I do not see an alternative to throwing something. Perhaps a new exception, with a better, more descriptive message?